### ISSUE TITLE
Follow Up Flags

### ISSUE THREAD

**Issue by fruitloop** [ecederstrand/exchangelib#324]
Hi. I use Office365 and I use the follow up flags and categories a lot. 

Since the Exchange mail rules are not that helpful I want to use your library to:
- set flags (with start and end date)
- set categories
- mark the message as read
- move the message to a subfolder

I guess the filtering of mails should not be a problem. Your README seems rather informative on that subject. And since my script would run a lot I could probably filter to only the last N minutes which should speed up the filtering.

To my actual question:
Is there a way to set the flags? With start and due date. Also: in my App on Android (Nine folders) I can change the subject of these "Tasks" 
I guess it should be possible to modify the subject of one flagged email and look for changes to reproduce this behaviour. However I have not found any hint that an email is indeed flagged. They definitely don't show up in the tasks. At lest not the last time I looked.

I hope this makes sense.
Happy new year and keep up the good work.

Best regards
Roland

**Comment by ecederstrand** [comment]
The follow-up flag is not part of the standard fields in EWS, but you can access it using extended properties. See https://github.com/ecederstrand/exchangelib/issues/85 on how to do that. I have no idea how to set start- and end dates on the flag - maybe it's also stored as extended properties.

**Comment by ecederstrand** [comment]
The remaining three can be achieved with something like:

```python
item.categories = ['foo', 'bar']
item.is_read = True
item.save()
item.move(to_folder=account.root / 'some_folder' / 'of' / 'your_choice')
```

**Comment by fruitloop** [comment]
thanks for the quick reply, setting flags works fine as described in #85.
However they are set without start/due date.

I found this site (https://social.msdn.microsoft.com/Forums/office/en-US/8fccfcf9-e7ea-43b2-821d-2fb611855428/ews-managed-api-setting-followup-flags-consistent-with-outlook?forum=exchangesvrdevelopment) stating that the due date should be `0x8105`

Therefore I tried this:
```
class FlagDue(ExtendedProperty):
  property_tag = 0x8105
  property_type = 'EWSDateTime'
Message.register('flag_due', FlagDue)
```

which gives me that:
```
Traceback (most recent call last):
  File "mail.py", line 16, in <module>
    Message.register('flag_due', FlagDue)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/items.py", line 376, in register
    attr_cls.validate_cls()
  File "/usr/local/lib/python3.6/site-packages/exchangelib/extended_properties.py", line 111, in validate_cls
    "'property_tag' value '%s' is reserved for custom properties" % cls.property_tag_as_hex()
ValueError: 'property_tag' value '0x8105' is reserved for custom properties
```
did I do something wrong? Is there a way to access these **custom properties**?

**Comment by ecederstrand** [comment]
That error comes from the restrictions on PropertyTag as stated in https://msdn.microsoft.com/en-us/library/office/aa564843(v=exchg.150).aspx

You need to use the equivalent `PropertySetId` attribute and the `PropertyName` or `PropertyId` attribute for this property. Maybe something like this will work:

```python
class FlagDue(ExtendedProperty):
    property_set_id = 'Task'
    property_id = 0x8105
    property_type = 'EWSDateTime'
```

The docs are a bit vague on this, but it's possible you need to use a distinguished set instead:

```python
ExtendedProperty.DISTINGUISHED_SETS.add('Task')


class FlagDue(ExtendedProperty):
    distinguished_property_set_id = 'Task'
    property_id = 0x8105
    property_type = 'EWSDateTime'
```

If the latter works, then we need to permanently add `Task` to `DISTINGUISHED_SETS`.

**Comment by fruitloop** [comment]
I tried this:
```
class FlagDue(ExtendedProperty):
  property_set_id = 'Task'
  property_id = 0x8105
Message.register('flag_due', FlagDue)
```
which got me that:
```
Traceback (most recent call last):
  File "mail.py", line 23, in <module>
    Message.register('flag_due', FlagDue)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/items.py", line 376, in register
    attr_cls.validate_cls()
  File "/usr/local/lib/python3.6/site-packages/exchangelib/extended_properties.py", line 129, in validate_cls
    "'property_type' value '%s' must be one of %s" % (cls.property_type, sorted(cls.PROPERTY_TYPES))
ValueError: 'property_type' value 'None' must be one of ['ApplicationTime', 'Binary', 'BinaryArray', 'Boolean', 'CLSID', 'CLSIDArray', 'Currency', 'CurrencyArray', 'Double', 'DoubleArray', 'Float', 'FloatArray', 'Integer', 'IntegerArray', 'Long', 'LongArray', 'Short', 'ShortArray', 'String', 'StringArray']
```
since that worked I tried every value type the error message presented me with

resulting in a message like this:
```
EWS https://outlook.office365.com/EWS/Exchange.asmx, account roland@planitz.at: Exception in _get_elements: Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 76, in _get_elements
    response = self._get_response_xml(payload=payload)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 148, in _get_response_xml
    res = self._get_soap_payload(soap_response=soap_response_payload)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 192, in _get_soap_payload
    cls._raise_soap_errors(fault=fault)  # Will throw SOAPError or custom EWS error
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 215, in _raise_soap_errors
    raise vars(errors)[code](msg)
exchangelib.errors.ErrorInvalidRequest: The request is invalid.

Traceback (most recent call last):
  File "mail.py", line 32, in <module>
    for email in account.inbox.all():
  File "/usr/local/lib/python3.6/site-packages/exchangelib/queryset.py", line 197, in __iter__
    for val in result_formatter(self._query()):
  File "/usr/local/lib/python3.6/site-packages/exchangelib/queryset.py", line 272, in _as_items
    for i in iterable:
  File "/usr/local/lib/python3.6/site-packages/exchangelib/account.py", line 534, in fetch
    for i in GetItem(account=self).call(items=ids, additional_fields=additional_fields, shape=IdOnly):
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 454, in _pool_requests
    elems = r.get()
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/pool.py", line 608, in get
    raise self._value
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 434, in <lambda>
    lambda c: self._get_elements(payload=payload_func(c, **kwargs)),
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 76, in _get_elements
    response = self._get_response_xml(payload=payload)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 148, in _get_response_xml
    res = self._get_soap_payload(soap_response=soap_response_payload)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 192, in _get_soap_payload
    cls._raise_soap_errors(fault=fault)  # Will throw SOAPError or custom EWS error
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 215, in _raise_soap_errors
    raise vars(errors)[code](msg)
exchangelib.errors.ErrorInvalidRequest: The request is invalid.
```
for `ApplicationTime`, `Binary`, `BinaryArray`, `Boolean`, `CLSID`, `CLSIDArray`, `Currency`, `CurrencyArray`, `Double`, `DoubleArray`, `Float`, `FloatArray`, `Integer`, `IntegerArray`, `Long`, `LongArray`, `Short`, `ShortArray`, `String`, `StringArray`

so I tried this
```
ExtendedProperty.DISTINGUISHED_SETS.add('Task')
class FlagDue(ExtendedProperty):
  distinguished_property_set_id = 'Task'
  property_id = 0x8105

Message.register('flag_due', FlagDue)
```
which got me that:
```
Traceback (most recent call last):
  File "mail.py", line 23, in <module>
    Message.register('flag_due', FlagDue)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/items.py", line 376, in register
    attr_cls.validate_cls()
  File "/usr/local/lib/python3.6/site-packages/exchangelib/extended_properties.py", line 129, in validate_cls
    "'property_type' value '%s' must be one of %s" % (cls.property_type, sorted(cls.PROPERTY_TYPES))
ValueError: 'property_type' value 'None' must be one of ['ApplicationTime', 'Binary', 'BinaryArray', 'Boolean', 'CLSID', 'CLSIDArray', 'Currency', 'CurrencyArray', 'Double', 'DoubleArray', 'Float', 'FloatArray', 'Integer', 'IntegerArray', 'Long', 'LongArray', 'Short', 'ShortArray', 'String', 'StringArray']
```

Again I tried every property_type the error message gave me, giving me these errors:
```
Traceback (most recent call last):
  File "mail.py", line 32, in <module>
    for email in account.inbox.all():
  File "/usr/local/lib/python3.6/site-packages/exchangelib/queryset.py", line 197, in __iter__
    for val in result_formatter(self._query()):
  File "/usr/local/lib/python3.6/site-packages/exchangelib/queryset.py", line 272, in _as_items
    for i in iterable:
  File "/usr/local/lib/python3.6/site-packages/exchangelib/account.py", line 534, in fetch
    for i in GetItem(account=self).call(items=ids, additional_fields=additional_fields, shape=IdOnly):
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 456, in _pool_requests
    for elem in elems:
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 283, in _get_elements_in_response
    container_or_exc = self._get_element_container(message=msg, name=self.element_container_name)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 256, in _get_element_container
    self._raise_errors(code=response_code, text=msg_text, msg_xml=msg_xml)
  File "/usr/local/lib/python3.6/site-packages/exchangelib/services.py", line 273, in _raise_errors
    raise vars(errors)[code](text)
exchangelib.errors.ErrorInvalidExtendedProperty: The extended property attribute combination is invalid. (field: <ns0:ExtendedFieldURI xmlns:ns0="http://schemas.microsoft.com/exchange/services/2006/types" DistinguishedPropertySetId="Task" PropertyId="33029" PropertyType="ApplicationTime" />)
```

Same for `Binary`, `BinaryArray`, `Boolean`, `CLSID`, `CLSIDArray`, `Currency`, `CurrencyArray`,  `Double`, `DoubleArray`, `Float`, `FloatArray`, `Integer`, `IntegerArray`, `Long`, `LongArray`, `Short`, `ShortArray`, `String`, `StringArray`

I have no clue what to try next.

**Comment by fruitloop** [comment]
I did find this though:
https://msdn.microsoft.com/en-us/library/office/jj191179(v=exchg.150).aspx

does that help?

**Comment by ecederstrand** [comment]
Hmm, I'll have to dig deeper when I have time. Thanks for the link to EWS docs. There's an apparent inconsistency there, because `Item` is listed as a parent of `Flag`, but when you go to the documentation for `Item`, then `Flag` isn't listed as a child element. Oh well.

**Comment by ecederstrand** [comment]
Ok, I actually got this working now. It was a bit convoluted.

First, the due date is stored in the [`PidLidTaskDueDate`](https://msdn.microsoft.com/en-us/library/office/cc839641.aspx) property. The property ID is `0x00008105` and the property set is `PSETID_Task`. But EWS wants the UUID for `PSETID_Task`, so we look that up in the [MS-OXPROPS](https://msdn.microsoft.com/en-us/library/cc433490(v=exchg.80).aspx) pdf. The UUID is `00062003-0000-0000-C000-000000000046`. Finally, we need to un-comment the `SystemTime` property type in the code, so we can use it. See the commit ba71e21. exchangelib will automatically de-serialize it to an `EWSDateTime`.

In conclusion, the definition for the due date becomes:

```python
class FlagDue(ExtendedProperty):
    property_set_id = '00062003-0000-0000-C000-000000000046'
    property_id = 0x8105
    property_type = 'SystemTime'

Message.register('flag_due', FlagDue)

for i in a.inbox.all():
    print(i.flag_due)
```

**Comment by fruitloop** [comment]
Great, that works for me!

just for documentation here's what I did:
```
class Flag(ExtendedProperty):
    property_tag = 0x1090
    property_type = 'Integer'
Message.register('flag', Flag)

class FlagDue(ExtendedProperty):
    property_set_id = '00062003-0000-0000-C000-000000000046'
    property_id = 0x8105
    property_type = 'SystemTime'
Message.register('flag_due', FlagDue)

class FlagStart(ExtendedProperty):
    property_set_id = '00062003-0000-0000-C000-000000000046'
    property_id = 0x8104
    property_type = 'SystemTime'
Message.register('flag_start', FlagStart)

for email in account.inbox.all():
    email.flag = 2
    email.flag_start = tz.localize(EWSDateTime(2018, 1, 27))
    email.flag_due = tz.localize(EWSDateTime(2018, 1, 28))
    email.save()
```

**Comment by krwester** [comment]
When I use the code 

```
class FlagDue(ExtendedProperty):
    property_set_id = '00062003-0000-0000-C000-000000000046'
    property_id = 0x8105
    property_type = 'SystemTime'

Message.register('flag_due', FlagDue)

for i in a.inbox.all():
    print(i.flag_due)
```
I get "None" for all messages, regardless of the date associated with the followup flag.   Any hints on troubleshooting?

**Comment by ecederstrand** [comment]
Are you sure these items have the flag set?

I think there are some (Windows-only) tools to inspect extended properties on items in an Exchange store. You could try that to find out where the flag values are registered on your items.

**Comment by krwester** [comment]
The flag is definitely set -- I just cant figure out how to get the flag
due date.

**Comment by ecederstrand** [comment]
Maybe your flags are registered under a different property ID? You can check in the request XML that we are actually requesting the field, as specified. See https://github.com/ecederstrand/exchangelib#troubleshooting

**Comment by fruitloop** [comment]
LG Roland
________________________________

**Comment by krwester** [comment]
From what I can tell, the field must be registered under something else.
 It's not showing up in the XML.    Is there any way to get the full list
of extended properties?

**Comment by krwester** [comment]
Ok some progress -- this is against exchange 2010_SP2.    If I run:

```
class FlagDue(ExtendedProperty):
    property_set_id = '00062003-0000-0000-C000-000000000046'
    property_id = 33029
    property_type = 'SystemTime'

Message.register('flag_due', FlagDue)
```

Then I can get the following in the XML log, but I can't figure out how to get that value into a variable.

```
<t:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyID="33029" PropertyType="SystemTime"/>
<t:Value>2019-08-20T14:00:00-10</t:Value>
```
Which is the UTC value and the offset of the due date on the followup flag.

**Comment by ecederstrand** [comment]
Ok, so it looks like you send a request for the field in one format, and EWS returns the field in another format, so we never match the field in the request with the field in the response. Notice the PropertySetId/DistinguishedPropertySetId difference. We only match request and response fields if all property values match. See https://github.com/ecederstrand/exchangelib/blob/master/exchangelib/fields.py#L1215

Try setting up the field like this instead:

```python
class FlagDue(ExtendedProperty):
    distinguished_property_set_id = 'Task'
    property_id = 33029
    property_type = 'SystemTime'
```

**Comment by ecederstrand** [comment]
We could probably loosen the exact match by implementing this translation table to match `property_set_id` and `distinguished_property_set_id` values: https://docs.microsoft.com/en-us/office/client-developer/outlook/mapi/commonly-used-property-sets

**Comment by ecederstrand** [comment]
Hi @krwester 

Can you test the referenced patch? Is should allow us to survive the situation above where you request `property_set_id = '00062003-0000-0000-C000-000000000046'` and the server returns `distinguished_property_set_id = 'Task'`.

**Comment by ecederstrand** [comment]
What error does it give?

**Comment by ecederstrand** [comment]
Pushed a new commit to allow `Task` as a `distinguished_property_set_id` value.

**Comment by krwester** [comment]
The following setup, with the latest commit gives the anticipated results:
```
class FlagDue(ExtendedProperty):
    distinguished_property_set_id = 'Task'
    property_id = 0x8105
    property_type = 'SystemTime'
```

**Comment by ecederstrand** [comment]
Great! Thanks for the test.
